bears/general/ArchiveOrgBear.py

```diff
@@ -0,0 +1,111 @@
+import requests
+
+from bears.general.InvalidLinkBear import InvalidLinkBear
+
+from coalib.settings.Setting import typed_dict
+from coalib.settings.Setting import typed_list
+from coalib.results.Result import Result
+from coalib.results.RESULT_SEVERITY import RESULT_SEVERITY
+
+from dependency_management.requirements.PipRequirement import PipRequirement
+
+from memento_client import MementoClient
+
+from urllib.parse import urlparse
+
+
+class ArchiveOrgBear(InvalidLinkBear):
+    DEFAULT_TIMEOUT = 15
+    LANGUAGES = {'All'}
+    REQUIREMENTS = {PipRequirement('memento_client', '0.5.3')}
+    AUTHORS = {'The coala developers'}
+    AUTHORS_EMAILS = {'coala-devel@googlegroups.com'}
+    LICENSE = 'AGPL-3.0'
+    CAN_DETECT = {'Documentation'}
+    DEFAULT_IGNORE = [
+        'http://web.archive.org/**',
+    ]
+
+    def analyze_links_in_file(self, file, network_timeout, link_ignore_regex,
+                              link_ignore_list):
+        for link, line_number, link_context in self.extract_links_from_file(
+                file, link_ignore_regex, link_ignore_list):
+
+            host = urlparse(link).netloc
+            code = InvalidLinkBear.get_status_code(
+                link,
+                network_timeout.get(host)
+                if host in network_timeout
+                else network_timeout.get('*')
+                if '*' in network_timeout
+                else self.DEFAULT_TIMEOUT)
+            if code and 200 <= code < 400:
+                yield line_number + 1, link, code, link_context
+
+    @staticmethod
+    def check_archive(mc, link):
+        try:
+            mc.get_memento_info(link)['mementos']
+        except KeyError:
+            return False
+        return True
+
+    @staticmethod
+    def get_redirect_urls(link):
+        urls = []
+        request_link = link
+
+        redirect_finished = False
+        while not redirect_finished:
+            resp = requests.head(request_link)
+            try:
+                next_url = resp.headers['Location']
+                urls.append(next_url)
+                request_link = next_url
+            except KeyError:
+                redirect_finished = True
+
+        return urls
+
+    def run(self, filename, file,
+            network_timeout: typed_dict(str, int, DEFAULT_TIMEOUT)=dict(),
+            link_ignore_regex: str='([.\/]example\.com|\{|\$)',
+            link_ignore_list: typed_list(str)=DEFAULT_IGNORE,
+            follow_redirects: bool=True):
```

Comment:
the documentation should also describe how this is useful because it's not fully trivial IMO